chia/full_node/subscriptions.py

```diff
@@ -0,0 +1,133 @@
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass, field
+from typing import Dict, List, Set
+
+from chia.types.blockchain_format.sized_bytes import bytes32
+
+log = logging.getLogger(__name__)
+
+
+# The PeerSubscriptions class is essentially a multi-index container. It can be
+# indexed by peer_id, coin_id and puzzle_hash.
+@dataclass(frozen=True)
+class PeerSubscriptions:
+    # TODO: use NewType all over to describe these various uses of the same types
+    # Puzzle Hash : Set[Peer ID]
+    _coin_subscriptions: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
+    # Puzzle Hash : Set[Peer ID]
+    _ph_subscriptions: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
+    # Peer ID: Set[Coin ids]
+    _peer_coin_ids: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
+    # Peer ID: Set[puzzle_hash]
+    _peer_puzzle_hash: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
+    # Peer ID: subscription count
+    _peer_sub_counter: Dict[bytes32, int] = field(default_factory=dict, init=False)
+
+    def has_ph_subscription(self, ph: bytes32) -> bool:
+        return ph in self._ph_subscriptions
+
+    def has_coin_subscription(self, coin_id: bytes32) -> bool:
+        return coin_id in self._coin_subscriptions
+
+    def add_ph_subscriptions(self, peer_id: bytes32, phs: List[bytes32], max_items: int) -> None:
+        puzzle_hash_peers = self._peer_puzzle_hash.setdefault(peer_id, set())
+        existing_sub_count = self._peer_sub_counter.setdefault(peer_id, 0)
+
+        # if we've reached the limit on number of subscriptions, just bail
+        if existing_sub_count >= max_items:
+            log.info(
+                "peer_id: %s reached max number of puzzle-hash subscriptions. "
+                "Not all its coin states will be reported",
+                peer_id,
+            )
+            return
```

Comment:
maybe return `False` or `True` so function caller could potentially report to the peer that not all subscriptions were processed.
Not sure how else will wallet/peer know this.

Reply:
right, it would be nice to have some feedback apart from logging. The wallet protocol doesn't really support that though, so that's a bigger change